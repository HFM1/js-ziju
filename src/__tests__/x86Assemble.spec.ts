import {
  Assembler
} from '../index'
import { build, readSource } from './utils'

describe('compiler x86', () => {
  it('functionCall', () => {
    let result = new Assembler(readSource('tests/functionCall.js')).compile()

    expect(result.assembly).toMatchSnapshot()

    build(result.assembly,'functionCall')
  });
});
