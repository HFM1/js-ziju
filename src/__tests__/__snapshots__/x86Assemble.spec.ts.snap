// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler x86 functionCall 1`] = `
"  .global _main

  .text

printChar:
  PUSH RBP
  MOV RBP, RSP

  PUSH 1
  MOV RAX, RBP
  ADD RAX, 16 # $$c
  PUSH RAX
  PUSH 1
  POP RDX
  POP RSI
  POP RDI
  MOV RAX, 33554436
  SYSCALL
  PUSH RAX

  POP RAX
  POP RBP

  RET

print:
  PUSH RBP
  MOV RBP, RSP

    # ADD
    PUSH 48
      # DIV
      PUSH [RBP + 16] # n
      PUSH 10
      POP RAX
      XCHG [RSP], RAX
      XOR RDX, RDX
      DIV QWORD PTR [RSP]
      MOV [RSP], RDX
    POP RAX
    ADD [RSP], RAX
    # End ADD
  CALL printChar
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

add:
  PUSH RBP
  MOV RBP, RSP

    # ADD
    PUSH [RBP + 24] # a
    PUSH [RBP + 16] # b
    POP RAX
    ADD [RSP], RAX
    # End ADD
  POP RAX
  POP RBP

  RET

main:
  PUSH RBP
  MOV RBP, RSP

  PUSH 2
  PUSH 3
  CALL add
  ADD RSP, 16
  PUSH RAX

  CALL print
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

_main:
  CALL main
  MOV RDI, RAX
  MOV RAX, 33554433
  SYSCALL
"
`;

exports[`compiler x86 print 1`] = `
"  .global _main

  .text

printChar:
  PUSH RBP
  MOV RBP, RSP

  PUSH 1
  MOV RAX, RBP
  ADD RAX, 16 # $$c
  PUSH RAX
  PUSH 1
  POP RDX
  POP RSI
  POP RDI
  MOV RAX, 33554436
  SYSCALL
  PUSH RAX

  POP RAX
  POP RBP

  RET

print:
  PUSH RBP
  MOV RBP, RSP

    # ADD
    PUSH 48
      # DIV
      PUSH [RBP + 16] # n
      PUSH 10
      POP RAX
      XCHG [RSP], RAX
      XOR RDX, RDX
      DIV QWORD PTR [RSP]
      MOV [RSP], RDX
    POP RAX
    ADD [RSP], RAX
    # End ADD
  CALL printChar
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

main:
  PUSH RBP
  MOV RBP, RSP

  PUSH 3
  CALL print
  MOV [RSP], RAX

  POP RAX
  POP RBP

  RET

_main:
  CALL main
  MOV RDI, RAX
  MOV RAX, 33554433
  SYSCALL
"
`;
